### PR TITLE
Completely remove raw pointers from AppFactory

### PR DESCRIPTION
**Pull request by permcody**
refs #7714

### CODE REVIEW

**Comment by permcody** [comment]
@andrsd - Will you take a look and hopefully merge. This PR fails Tiamat, but I have a build working on my desktop now so I've fixed the deprecated API. The tests don't pass on my local system but I think that's just because these tests are so brittle. Once this is merged, I'll push the patch to test against Tiamat (which has a MOOSE submodule) and see if I can get it working.

File: unit/include/PorousFlowBrineCO2Test.h [hunk 1]
```diff
@@ -93,7 +93,7 @@ class PorousFlowBrineCO2Test : public ::testing::Test
     _fp = &_fe_problem->getUserObject<PorousFlowBrineCO2>("fp");
   }
 
-  std::unique_ptr<MooseApp> _app;
+  std::shared_ptr<MooseApp> _app;
```
Comment (andrsd) [inline unit/include/PorousFlowBrineCO2Test.h]: `MooseAppPtr`?

File: unit/include/PorousFlowFluidStateFlashTest.h [hunk 2]
```diff
@@ -59,7 +59,7 @@ class PorousFlowFluidStateFlashTest : public ::testing::Test
     _fp = &_fe_problem->getUserObject<PorousFlowFluidStateFlash>("fp");
   }
 
-  std::unique_ptr<MooseApp> _app;
+  std::shared_ptr<MooseApp> _app;
```
Comment (andrsd) [inline unit/include/PorousFlowFluidStateFlashTest.h]: `MooseAppPtr`

File: unit/include/PorousFlowWaterNCGTest.h [hunk 3]
```diff
@@ -85,7 +85,7 @@ class PorousFlowWaterNCGTest : public ::testing::Test
     _fp = &_fe_problem->getUserObject<PorousFlowWaterNCG>("fp");
   }
 
-  std::unique_ptr<MooseApp> _app;
+  std::shared_ptr<MooseApp> _app;
```
Comment (andrsd) [inline unit/include/PorousFlowWaterNCGTest.h]: `MooseAppPtr`

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by permcody** [comment]
This should be ready for merge. I'll fixup Tiamat-mock when this is in.

**Review by andrsd — APPROVED** [review]

**Pull request merged**